[CGUIDialogContextMenu] fix memory leak

repojohnray opened the pull request:
## Description
This memory leak is happening when a context menu is opened. The logs are available below.

## Motivation and context
The goal is to clean up some issues detected by the sanitizer.

## How has this been tested?
The option "-fsanitize=leak" is required and should be passed to the compiler and the linker. For a fully sanitized executable, you could use the following options: "-fsanitize=bounds -fsanitize=address -fsanitize=undefined -fsanitize=unreachable -fsanitize=bool -fsanitize=enum -fsanitize=leak -fsanitize-recover=address"

An example below, or you could use the cmake option: -DECM_ENABLE_SANITIZERS
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS_DEBUG="-ggdb -O2 -fno-omit-frame-pointer -fsanitize=leak" -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=leak" -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=leak"

The gcc sanitizer operates properly with a debug version of Python-3.11 compiled with the following options: '--without-pymalloc --with-pydebug --with-valgrind'. As a reminder with Python-3.8 the same options were not enough to generate a "sanitized" compatible python library.

## What is the effect on users?

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

repojohnray commented:
```
Indirect leak of 116352 byte(s) in 72 object(s) allocated from:
    #0 0x7fd1edf94f37 in operator new(unsigned long) (/usr/lib64/libasan.so.6+0xb2f37)
    #1 0x75ea379 in CGUIDialogContextMenu::SetupButtons() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:131
    #2 0x75f925c in CGUIDialogContextMenu::Show(CContextButtons const&, int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:658
    #3 0x493675d in CGUIMediaWindow::OnPopupMenu(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:1784
    #4 0x4995b6e in CGUIMediaWindow::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:316
    #5 0x399289d in CGUIWindowVideoBase::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:205
    #6 0x39b742f in CGUIWindowVideoNav::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:243
    #7 0x6859bc3 in CGUIBaseContainer::OnClick(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:885
    #8 0x6884262 in CGUIBaseContainer::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:462
    #9 0x719062e in CGUIWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindow.cpp:423
    #10 0x494994b in CGUIMediaWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:191
    #11 0x39547fa in CGUIWindowVideoBase::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:108
    #12 0x39b576a in CGUIWindowVideoNav::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:93
    #13 0x7234aa1 in CGUIWindowManager::HandleAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1188
    #14 0x7235c56 in CGUIWindowManager::OnAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1133
    #15 0x7bd113c in CApplication::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:997
    #16 0x62c0abe in CInputManager::HandleKey(CKey const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:660
    #17 0x62d2417 in CInputManager::OnEvent(XBMC_Event&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:447
    #18 0x7be181e in CApplication::HandlePortEvents() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:337
    #19 0x7be498c in CApplication::FrameMove(bool, bool) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1808
    #20 0x7b93798 in CApplication::Run() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1912
    #21 0x5d09d04 in XBMC_Run xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/xbmc.cpp:61
    #22 0x113d99a in main xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/posix/main.cpp:71
    #23 0x7fd1c0633236 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #24 0x7fd1c06332f1 in __libc_start_main_impl ../csu/libc-start.c:392
    #25 0x1262b70 in _start (/usr/local/lib64/kodi/kodi.bin+0x1262b70)

Indirect leak of 33984 byte(s) in 72 object(s) allocated from:
    #0 0x7fd1edf94f37 in operator new(unsigned long) (/usr/lib64/libasan.so.6+0xb2f37)
    #1 0x74f2b2b in CGUITextureGL::Clone() const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUITextureGL.cpp:43
    #2 0x68ee6bd in CGUIButtonControl::CGUIButtonControl(CGUIButtonControl const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIButtonControl.cpp:46
    #3 0x75ea3aa in CGUIDialogContextMenu::SetupButtons() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:131
    #4 0x75f925c in CGUIDialogContextMenu::Show(CContextButtons const&, int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:658
    #5 0x493675d in CGUIMediaWindow::OnPopupMenu(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:1784
    #6 0x4995b6e in CGUIMediaWindow::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:316
    #7 0x399289d in CGUIWindowVideoBase::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:205
    #8 0x39b742f in CGUIWindowVideoNav::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:243
    #9 0x6859bc3 in CGUIBaseContainer::OnClick(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:885
    #10 0x6884262 in CGUIBaseContainer::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:462
    #11 0x719062e in CGUIWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindow.cpp:423
    #12 0x494994b in CGUIMediaWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:191
    #13 0x39547fa in CGUIWindowVideoBase::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:108
    #14 0x39b576a in CGUIWindowVideoNav::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:93
    #15 0x7234aa1 in CGUIWindowManager::HandleAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1188
    #16 0x7235c56 in CGUIWindowManager::OnAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1133
    #17 0x7bd113c in CApplication::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:997
    #18 0x62c0abe in CInputManager::HandleKey(CKey const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:660
    #19 0x62d2417 in CInputManager::OnEvent(XBMC_Event&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:447
    #20 0x7be181e in CApplication::HandlePortEvents() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:337
    #21 0x7be498c in CApplication::FrameMove(bool, bool) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1808
    #22 0x7b93798 in CApplication::Run() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1912
    #23 0x5d09d04 in XBMC_Run xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/xbmc.cpp:61
    #24 0x113d99a in main xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/posix/main.cpp:71
    #25 0x7fd1c0633236 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #26 0x7fd1c06332f1 in __libc_start_main_impl ../csu/libc-start.c:392
    #27 0x1262b70 in _start (/usr/local/lib64/kodi/kodi.bin+0x1262b70)

Indirect leak of 33984 byte(s) in 72 object(s) allocated from:
    #0 0x7fd1edf94f37 in operator new(unsigned long) (/usr/lib64/libasan.so.6+0xb2f37)
    #1 0x74f2b2b in CGUITextureGL::Clone() const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUITextureGL.cpp:43
    #2 0x68ee38d in CGUIButtonControl::CGUIButtonControl(CGUIButtonControl const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIButtonControl.cpp:45
    #3 0x75ea3aa in CGUIDialogContextMenu::SetupButtons() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:131
    #4 0x75f925c in CGUIDialogContextMenu::Show(CContextButtons const&, int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/dialogs/GUIDialogContextMenu.cpp:658
    #5 0x493675d in CGUIMediaWindow::OnPopupMenu(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:1784
    #6 0x4995b6e in CGUIMediaWindow::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:316
    #7 0x399289d in CGUIWindowVideoBase::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:205
    #8 0x39b742f in CGUIWindowVideoNav::OnMessage(CGUIMessage&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:243
    #9 0x6859bc3 in CGUIBaseContainer::OnClick(int) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:885
    #10 0x6884262 in CGUIBaseContainer::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIBaseContainer.cpp:462
    #11 0x719062e in CGUIWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindow.cpp:423
    #12 0x494994b in CGUIMediaWindow::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/windows/GUIMediaWindow.cpp:191
    #13 0x39547fa in CGUIWindowVideoBase::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoBase.cpp:108
    #14 0x39b576a in CGUIWindowVideoNav::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/video/windows/GUIWindowVideoNav.cpp:93
    #15 0x7234aa1 in CGUIWindowManager::HandleAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1188
    #16 0x7235c56 in CGUIWindowManager::OnAction(CAction const&) const xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/guilib/GUIWindowManager.cpp:1133
    #17 0x7bd113c in CApplication::OnAction(CAction const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:997
    #18 0x62c0abe in CInputManager::HandleKey(CKey const&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:660
    #19 0x62d2417 in CInputManager::OnEvent(XBMC_Event&) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/input/InputManager.cpp:447
    #20 0x7be181e in CApplication::HandlePortEvents() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:337
    #21 0x7be498c in CApplication::FrameMove(bool, bool) xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1808
    #22 0x7b93798 in CApplication::Run() xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/application/Application.cpp:1912
    #23 0x5d09d04 in XBMC_Run xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/xbmc.cpp:61
    #24 0x113d99a in main xbmc-b86192a01b3acb988c6f60899127cb4d10a82f7b/xbmc/platform/posix/main.cpp:71
    #25 0x7fd1c0633236 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #26 0x7fd1c06332f1 in __libc_start_main_impl ../csu/libc-start.c:392
    #27 0x1262b70 in _start (/usr/local/lib64/kodi/kodi.bin+0x1262b70)

```

CastagnaIT commented on xbmc/dialogs/GUIDialogContextMenu.cpp:
```diff
@@ -126,17 +126,16 @@ void CGUIDialogContextMenu::SetupButtons()
-      pButton->SetLabel(m_buttons[i].second);
-      if (pGroupList)
-      {
+  if (pGroupList)
```
please add brackets for "if" statement

repojohnray replied:
@CastagnaIT, the commit is updated.

repojohnray committed:
[CGUIDialogContextMenu] fix memory leak

This memory leak is happening when a context menu is opened.

This change addresses two issues:
- Missing delete.
- Group list not defined (the behavior is changed).

CastagnaIT reviewed (approved):
please do also backport

ksooo reviewed (commented):
If `SetupButtons` gets called  multiple times with no `OnDeinitWindow` in between, then we still have the memory leak. `SetupButtons` should be extended to delete the buttons like done in `OnDeInitWindow` before creating new ones.

repojohnray commented:
@ksooo, is this possible to call a context menu multiple times with SetupButtons? If no, the memory leak would indicate an implementation mistake.

CrystalP commented:
That doesn't seem possible. The flow to SetupButton comes from the Show and ShowAndGetChoice and the caller of the context menu doesn't get control back and can't call them again until after OnDeinitWindow ran.
There is only one context menu instance and it's a dialog so it seems to me that the only possibility would be for a context menu item to invoke another context menu (a sub-context menu?) and I didn't see such code.

CrystalP reviewed (approved):
Looks ok runs ok although I'm not the best one for GUI stuff

CrystalP commented:
@ksooo care to update your comment? See my analysis a bit earlier.

@repojohnray  can you please clarify in the description that this fix addresses two leaks
- missing delete
- group list not defined - with behavior change

The behavior changed a bit when you moved the `if (pGroupList)`, but during testing it seems the created controls were not usable anyway so looks OK to skip their creation.
Actually it would be good to log something when a group list is not defined, with categories error / skinning I guess.

repojohnray commented:
@CrystalP, the description is updated at the commit level. Logging these events is a good idea.
